### PR TITLE
Added support for find externals to also find library-only packages

### PR DESCRIPTION
**Pull request by bvanessen**
Similar to the existing ability to find externals via a executable
binary, this allows a package to define a shared library object that
can be used to find a detectable version of an installed package.
Requires a `libraries = [lib<SHARED_LIB>.so]` field in the recipe as
well as the `determine_version` class function.

### CODE REVIEW

**Comment by bvanessen** [comment]
@scheibelp @becker33 @tgamblin Here is the PR to add the feature to find externals by shared library.

**Comment by adamjstewart** [comment]
This looks really cool! I wonder if we're starting to duplicate a lot of the functionality in Spack however. Packages already have `commands` and `libs` attributes to define the names of binaries and libraries. I wonder if we could reuse them during external package detection.

**Comment by adamjstewart** [comment]
My only other concern is that library filename versions often don't match package versions. OpenMPI is a good example of this, the `package.py` has a mapping between versions and library names.

**Comment by bvanessen** [comment]
> My only other concern is that library filename versions often don't match package versions. OpenMPI is a good example of this, the `package.py` has a mapping between versions and library names.

Absolutely, this will probably not work for all libraries. I will checkout OpenMPI, but since the package has to define a `determine_version` function it can be fairly flexible.

**Comment by bvanessen** [comment]
> This looks really cool! I wonder if we're starting to duplicate a lot of the functionality in Spack however. Packages already have `commands` and `libs` attributes to define the names of binaries and libraries. I wonder if we could reuse them during external package detection.

@adamjstewart How would I access the `libs` field that is already defined?

**Comment by adamjstewart** [comment]
@bvanessen you would do something like `self.spec['foo'].libs`. See https://spack.readthedocs.io/en/latest/llnl.util.html#llnl.util.filesystem.LibraryList for a full list of properties. We can modify this to add more features if that makes sense, or we can not use it at all. Just trying to prevent code duplication if possible.

**Review by srekolam — COMMENTED** [review]
looks good to me. I have reviewed the changes for hipblas, rccl. If other rocm recipes require similar change, do let me know . i will work on them, Thanks

**Comment by bvanessen** [comment]
> looks good to me. I have reviewed the changes for hipblas, rccl. If other rocm recipes require similar change, do let me know . i will work on them, Thanks

I think that most of the HIP/ROCm stack needs this to be implemented.  I just wanted to get a few examples in place first.

**Comment by bvanessen** [comment]
> @bvanessen you would do something like `self.spec['foo'].libs`. See https://spack.readthedocs.io/en/latest/llnl.util.html#llnl.util.filesystem.LibraryList for a full list of properties. We can modify this to add more features if that makes sense, or we can not use it at all. Just trying to prevent code duplication if possible.

So I am thinking that this could be a double edged sword.  Automatically detecting the available libraries and looking for them would simplify the burden on the package maintainers, but it would mean that Spack starts looking for many more packages than maybe are ready for the feature.  Right now there is a bit of balance, where it is a bit tedious to make sure that the version is properly extracted from the dynamic library object, but only packages where someone put in the effort can use the feature.

**Comment by adamjstewart** [comment]
I _think_ the way that things are currently implemented for executable detection is that we don't consider an external package "found" unless both the executable is found and a version is extracted. I actually object to this design and would prefer the possibility of having version-less executables (if a specific version is required by a package, then don't use it, else it should work fine). So basically, it might be fine to use `libs` for now, but I would prefer a situation where we didn't need to extract the version name every time, which would make it a bad idea to use `libs` as you've suggested. Although we've implemented `libs` for quite a few packages already, many of them don't work properly.

**Comment by bvanessen** [comment]
@adamjstewart @haampie @alalazo @srekolam @scheibelp What are the next steps for this PR?  I see that @srekolam left review comments, but I cannot actually see what they were.

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: Should "PATH" be "LD_LIBRARY_PATH" or "path_hints"? I wouldn't expect to find libraries in `PATH`.

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: I think `executables_in_path` should be checking `if is_executable` and `libraries_in_path` should be checking `is_readable`, but as written the changes do the opposite of that.

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: As part of completing this, we should either answer each of these questions or flesh them out:

> What tool will we use? Is that per-package?

It looks like in all cases you extract the library version from the filename. IIRC we've discussed packages which don't follow this pattern.

The most definitive method I can find for extracting the version from Linux libraries is invoking `strings`, in https://unix.stackexchange.com/questions/58846/viewing-linux-library-executable-version-info (according to that, the ELF data doesn't necessarily store any info about the version other than perhaps the major version number). 

> Call determine_spec_details on each package?

I think this is already answered since it's handled in the implementation, but to address this explicitly: this is called once for each unique prefix/package combination. The next point will complicate this though.

> Do we need separate determine_spec_details_exe and determine_spec_details_lib?

As of now it appears that you do not: each packages either defines `.libraries` or `.executables`, so `determine_spec_details` always knows whether it will be getting one or the other. I think ultimately it would be good to allow for packages which define both, so the method should be updated.

I think the most robust way to support this is to modify `determine_spec_details`: add a separate keyword parameter for libraries and for executables. I prefer this approach (e.g. compared to having separate `determine_spec_details_libraries` and `determine_spec_details_exes`) because it can use the library and exe information together to produce a spec (and could for example be used in the future to track external packages which do not have a common prefix for externals and libraries).

This would only work if `by_executable` and `by_library` were combined into one function; this would also mean that it is no-longer sensible for `determine_spec_details` to accept a single prefix: it would be called once for all associated exes/libraries and would be responsible for grouping by version (which isn't as difficult as might be implied at first, since the current implementation was already handling the possibility of having e.g. `gcc-7` and `gcc-8` executables in a combined prefix).

> How do we get compiler information?
> for things in /usr we can lookup by OS

Your suggestion seems good. It might be worth experimenting with `objdump` as suggested by https://unix.stackexchange.com/questions/719/can-we-get-compiler-information-from-an-elf-binary

> should enabling this be an assertion that the compiler doesn't matter?

One of the packages you updated here, `nccl`, appears to be C++, and IIRC ABI compatibility isn't guaranteed between different compilers for C++, so ideally we could figure out the compiler.

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: (minor: this `print` should be removed before merging)

**Review by scheibelp — COMMENTED** [review]
You have some questions in the implementation of `by_library`: I think answering these will take some time and discussion. I left a review comment which discusses each question in that function.

**Comment by bvanessen** [comment]
@scheibelp Are you happy with the changes that have been made?

File: lib/spack/spack/package.py
Comment (scheibelp) [inline lib/spack/spack/package.py:216]: "the directory containing the executables"

Need to revise that. Also:

"the executables and libraries that match the regex"

Can that be revised to clarify that it will not be a mix?

File: lib/spack/spack/package.py
Comment (scheibelp) [inline lib/spack/spack/package.py:194]: Can you raise an exception if `hasattr(cls, 'executables') and hasattr(cls, 'libraries')`?

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py:114]: Can you add a TODO comment here to consolidate this with `by_executable`? Namely:

* Packages should be able to define both .libraries and .executables in the future
* `determine_spec_details` should get all relevant libraries and executables in one call

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a few (hopefully minor) requests but other than that this looks good!

**Comment by nilsvu** [comment]
@bvanessen I think is will be very useful! I recently thought about this as well before coming across your PR, so I would like to share two thoughts with you:

1. You may be able to take some inspiration from CMake. It is already very good at finding libraries, and has years of experience with it. CMake's search procedure is documented here:  https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure. Going a step further, it can even be useful to look at CMake's individual find modules, such as [FindMPI](https://cmake.org/cmake/help/latest/module/FindMPI.html), [FindHDF5](https://cmake.org/cmake/help/latest/module/FindHDF5.html), [FindLAPACK](https://cmake.org/cmake/help/latest/module/FindLAPACK.html), [FindBoost](https://cmake.org/cmake/help/latest/module/FindBoost.html), etc. I opened https://github.com/spack/spack/issues/28570 recently to propose using these modules in `spack external find`. I did some basic tests (see e.g. https://github.com/nilsleiffischer/spack/blob/find_cmake/test_cmake/CMakeLists.txt) and I'm currently on the fence if this CMake-based approach will work. Either way, I think we can learn a lot from CMake here.
2. Please note that shared libraries do not always end in `.so`. I'm certainly no expert on all these C++ intrinsics, but on macOS I mostly see `.dylib` files, including in Spack installations.

**Comment by adamjstewart** [comment]
> Please note that shared libraries do not always end in `.so`. I'm certainly no expert on all these C++ intrinsics, but on macOS I mostly see `.dylib` files, including in Spack installations.

Spack already has a ton of built-in logic for finding libraries, we should really be reusing that. See `llnl.util.filesystem.find_libraries`. That correctly uses `.so` on Linux and `.dylib` on macOS. Although I've noticed a lot of packages that use `.so` on macOS too (especially Python) so we should probably just check for both.

**Comment by bvanessen** [comment]
@scheibelp I have updated the PR with your feedback.

**Comment by bvanessen** [comment]
@scheibelp @adamjstewart Is this PR still being held up by the windows support PR?

**Comment by scheibelp** [comment]
Yes: I'm still working on that PR.

**Comment by scheibelp** [comment]
@spackbot run pipeline

File: lib/spack/spack/detection/common.py
Comment (scheibelp) [inline lib/spack/spack/detection/common.py]: If there is no `lib`/`lib64` to strip, then we should probably either:

* Return the library_dir as the prefix (this is what `executable_prefix` does
* Raise an exception (which is also reasonable but might lead to additional changes to decide when/if to address the exception) 

In the second case, that would probably involve replacing:

```
                pkg_prefix = library_prefix(prefix)

                 if not pkg_prefix:
                     msg = "no lib/ or lib64/ dir found in {0}. Cannot "
                     "add it as a Spack package"
                     llnl.util.tty.debug(msg.format(prefix))
                     continue
``` 

with a try/catch.

File: lib/spack/spack/detection/common.py
Comment (scheibelp) [inline lib/spack/spack/detection/common.py]: I think `is_readable_file` is more appropriate (since this function excludes directories).

Also, this function isn't specifically related to detection, and so would likely be best off in `llnl.util.filesystem`.

File: lib/spack/spack/detection/path.py
Comment (scheibelp) [inline lib/spack/spack/detection/path.py]: You can remove this comment since it's redundant WRT the `by_executable` function.

File: lib/spack/spack/package.py
Comment (scheibelp) [inline lib/spack/spack/package.py]: Can you raise `spack.error.SpackError` here?

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a few (hopefully minor) requests but overall things look good - thanks!

File: var/spack/repos/builtin/packages/hipblas/package.py
Comment (nilsvu) [inline var/spack/repos/builtin/packages/hipblas/package.py:20]: I haven't read the code in detail, but I'd like to reiterate my warning from above again here: libraries don't always end with `.so`, on macOS it's mostly `.dylib`. Is this handled somewhere?

Comment (scheibelp) [reply var/spack/repos/builtin/packages/hipblas/package.py:20]: The strings placed in the `.libraries` attribute are treated like regular expressions and allow for substring matching. For example if you replaced `libhipblas.so` with `libhipblas`, and Mac OS had `libhipblas.dylib`, this would be able to find it.

I didn't want to require updating the regular expressions for specific packages in this PR: I thought that could occur in future PRs if users on those systems want to autodetect these packages there; in part I did this because I wasn't sure if the PR author had access to a Mac OS installation to confirm the name format for these packages.

Comment (bvanessen) [reply var/spack/repos/builtin/packages/hipblas/package.py:20]: @nilsvu Yes it would be great to add OX support, but that feels better in a separate PR so that this can merge.

**Comment by bvanessen** [comment]
@scheibelp I have resolved your code requests.

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
Thanks!

**Pull request merged**
